Exception when using Spring Cloud Stream

**Describe the bug**
When an application uses spring cloud stream and spring aspect, there will be an error when collect metrics.

**Steps to reproduce**
1. download the [demo](https://github.com/IcebergXTY/opentelemetry-cloud-stream-demo) code
2. change the rabbitmq address/username/password in application.yaml
3. start the application with agent
-javaagent:D:\\DevelopSoftware\\OpenTelemetry-agent\\opentelemetry-javaagent.jar
-Dotel.traces.exporter=logging
-Dotel.metrics.exporter=logging
-Dotel.logs.exporter=logging
4. after application started, wait for about two minutes, there will throw an exception look like this
```java
2022-11-17 16:47:45.519  INFO 35332 --- [(4)-10.60.52.17] o.s.web.servlet.DispatcherServlet        : Completed initialization in 1 ms
[otel.javaagent 2022-11-17 16:49:04:411 +0800] [PeriodicMetricReader-1] WARN io.opentelemetry.sdk.metrics.internal.state.CallbackRegistration - An exception occurred invoking callback for Instrument spring.integration.handlers.
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'luna-taskcore-exchange.taskcore.errors.handler': Initialization of bean failed; nested exception is java.lang.IllegalArgumentException: warning no match for this type name: com.example.demo.controller.CommonController [Xlint:invalidAbsoluteTypeName]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:628)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:673)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:661)
	at org.springframework.context.support.AbstractApplicationContext.getBeansOfType(AbstractApplicationContext.java:1300)
	at org.springframework.integration.config.IntegrationManagementConfigurer.lambda$registerComponentGauges$1(IntegrationManagementConfigurer.java:154)
	at io.opentelemetry.javaagent.shaded.instrumentation.micrometer.v1_5.DoubleMeasurementRecorder.accept(DoubleMeasurementRecorder.java:35)
	at io.opentelemetry.javaagent.shaded.instrumentation.micrometer.v1_5.DoubleMeasurementRecorder.accept(DoubleMeasurementRecorder.java:15)
	at io.opentelemetry.sdk.metrics.AbstractInstrumentBuilder.lambda$registerDoubleAsynchronousInstrument$0(AbstractInstrumentBuilder.java:87)
	at io.opentelemetry.sdk.metrics.internal.state.CallbackRegistration.invokeCallback(CallbackRegistration.java:97)
	at io.opentelemetry.sdk.metrics.internal.state.MeterSharedState.collectAll(MeterSharedState.java:96)
	at io.opentelemetry.sdk.metrics.SdkMeter.collectAll(SdkMeter.java:67)
	at io.opentelemetry.sdk.metrics.SdkMeterProvider$LeasedMetricProducer.collectAllMetrics(SdkMeterProvider.java:180)
	at io.opentelemetry.sdk.metrics.export.PeriodicMetricReader$Scheduled.doRun(PeriodicMetricReader.java:156)
	at io.opentelemetry.sdk.metrics.export.PeriodicMetricReader$Scheduled.run(PeriodicMetricReader.java:148)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: warning no match for this type name: com.example.demo.controller.CommonController [Xlint:invalidAbsoluteTypeName]
	at org.aspectj.weaver.tools.PointcutParser.parsePointcutExpression(PointcutParser.java:319)
	at org.springframework.aop.aspectj.AspectJExpressionPointcut.buildPointcutExpression(AspectJExpressionPointcut.java:227)
	at org.springframework.aop.aspectj.AspectJExpressionPointcut.obtainPointcutExpression(AspectJExpressionPointcut.java:198)
	at org.springframework.aop.aspectj.AspectJExpressionPointcut.getClassFilter(AspectJExpressionPointcut.java:177)
	at org.springframework.aop.support.AopUtils.canApply(AopUtils.java:226)
	at org.springframework.aop.support.AopUtils.canApply(AopUtils.java:289)
	at org.springframework.aop.support.AopUtils.findAdvisorsThatCanApply(AopUtils.java:321)
	at org.springframework.aop.framework.autoproxy.AbstractAdvisorAutoProxyCreator.findAdvisorsThatCanApply(AbstractAdvisorAutoProxyCreator.java:128)
	at org.springframework.aop.framework.autoproxy.AbstractAdvisorAutoProxyCreator.findEligibleAdvisors(AbstractAdvisorAutoProxyCreator.java:97)
	at org.springframework.aop.framework.autoproxy.AbstractAdvisorAutoProxyCreator.getAdvicesAndAdvisorsForBean(AbstractAdvisorAutoProxyCreator.java:78)
	at org.springframework.aop.framework.autoproxy.AbstractAutoProxyCreator.wrapIfNecessary(AbstractAutoProxyCreator.java:341)
	at org.springframework.aop.framework.autoproxy.AbstractAutoProxyCreator.postProcessAfterInitialization(AbstractAutoProxyCreator.java:293)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsAfterInitialization(AbstractAutowireCapableBeanFactory.java:455)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1808)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620)
	... 25 more
```

**What did you expect to see?**
No exception throw.

**What version are you using?**
Version 1.19.2

**Environment**
Compiler: Eclipse Temurin 1.8.0_302
OS: windows 10 21H1 64bit

**Additional context**
Anyway I think this problem is not caused by anyone, including spring and opentelemetry. It's just a coincidence.
I will try to analyze the it here.

---

The error is reported at `org.springframework.integration.config.IntegrationManagementConfigurer#registerComponentGauges`, As shown here
![image](https://user-images.githubusercontent.com/15325243/202594425-a63a27a9-bef6-4c70-89e0-ff41861757a8.png)
This part of the logic is lazy loading, only when called will get and initialization of type `org.springframework.messaging.MessageHandler`. Here will get three beanName, The error is raised by `luna-taskcore-exchange.taskcore.errors.handler`.
![image](https://user-images.githubusercontent.com/15325243/202594994-5dc1d74f-30ba-491a-9a47-111edd25d2ff.png)
When the `luna-taskcore-exchange.taskcore.errors.handler` bean is initializing, it's going to be processed by a BeanPostProcessor named `org.springframework.aop.aspectj.annotation.AnnotationAwareAspectJAutoProxyCreator`.
The `AnnotationAwareAspectJAutoProxyCreator` will try all the aspects in `org.springframework.aop.support.AopUtils#findAdvisorsThatCanApply` method if it is suitable for the handler bean.

On the other hand,  when spring try to load my `com.example.demo.aspect.ExceptionAspect`, first it will process the aspectj expression by `org.springframework.aop.aspectj.AspectJExpressionPointcut#obtainPointcutExpression`.Here, it will use the current thread's classloader to process the aspectj expression and this is AgentClassloader now.So when load my expression `execution(* com.example.demo.controller.CommonController.*(..))`, spring will raise an error as AgentClassloader don't know what CommonController is.
![image](https://user-images.githubusercontent.com/15325243/202596502-92ae2b96-1a5d-47d3-8a0f-1d8ebcf7e7e6.png)

---

The temporary solution is to load these bean in spring EventListener before the agent effect, but it's hard to know all the bean need to be early loaded. 
And as https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/7037 say, the agent actualy can't load the application class.
So there is any other solution to solve this problem?

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java
@@ -19,7 +19,6 @@ import io.opentelemetry.context.Scope;
 import io.opentelemetry.instrumentation.api.internal.ConfigPropertiesUtil;
 import io.opentelemetry.instrumentation.api.internal.EmbeddedInstrumentationProperties;
 import io.opentelemetry.javaagent.bootstrap.AgentClassLoader;
-import io.opentelemetry.javaagent.bootstrap.AgentInitializer;
 import io.opentelemetry.javaagent.bootstrap.BootstrapPackagePrefixesHolder;
 import io.opentelemetry.javaagent.bootstrap.ClassFileTransformerHolder;
 import io.opentelemetry.javaagent.bootstrap.DefineClassHelper;
@@ -78,7 +77,7 @@ public class AgentInstaller {
 
   private static final Map<String, List<Runnable>> CLASS_LOAD_CALLBACKS = new HashMap<>();
 
-  public static void installBytebuddyAgent(Instrumentation inst) {
+  public static void installBytebuddyAgent(Instrumentation inst, ClassLoader extensionClassLoader) {
     addByteBuddyRawSetting();
 
     Integer strictContextStressorMillis = Integer.getInteger(STRICT_CONTEXT_STRESSOR_MILLIS);
@@ -90,34 +89,37 @@ public class AgentInstaller {
     logVersionInfo();
     if (ConfigPropertiesUtil.getBoolean(JAVAAGENT_ENABLED_CONFIG, true)) {
       setupUnsafe(inst);
-      List<AgentListener> agentListeners = loadOrdered(AgentListener.class);
-      installBytebuddyAgent(inst, agentListeners);
+      List<AgentListener> agentListeners = loadOrdered(AgentListener.class, extensionClassLoader);
+      installBytebuddyAgent(inst, extensionClassLoader, agentListeners);
     } else {
       logger.fine("Tracing is disabled, not installing instrumentations.");
     }
   }
 
   private static void installBytebuddyAgent(
-      Instrumentation inst, Iterable<AgentListener> agentListeners) {
+      Instrumentation inst,
+      ClassLoader extensionClassLoader,
+      Iterable<AgentListener> agentListeners) {
 
     WeakRefAsyncOperationEndStrategies.initialize();
 
-    EmbeddedInstrumentationProperties.setPropertiesLoader(
-        AgentInitializer.getExtensionsClassLoader());
+    EmbeddedInstrumentationProperties.setPropertiesLoader(extensionClassLoader);
 
     setDefineClassHandler();
 
     // If noop OpenTelemetry is enabled, autoConfiguredSdk will be null and AgentListeners are not
     // called
-    AutoConfiguredOpenTelemetrySdk autoConfiguredSdk = installOpenTelemetrySdk();
+    AutoConfiguredOpenTelemetrySdk autoConfiguredSdk =
+        installOpenTelemetrySdk(extensionClassLoader);
 
     ConfigProperties sdkConfig = autoConfiguredSdk.getConfig();
     InstrumentationConfig.internalInitializeConfig(new ConfigPropertiesBridge(sdkConfig));
     copyNecessaryConfigToSystemProperties(sdkConfig);
 
-    setBootstrapPackages(sdkConfig);
+    setBootstrapPackages(sdkConfig, extensionClassLoader);
 
-    for (BeforeAgentListener agentListener : loadOrdered(BeforeAgentListener.class)) {
+    for (BeforeAgentListener agentListener :
+        loadOrdered(BeforeAgentListener.class, extensionClassLoader)) {
       agentListener.beforeAgent(autoConfiguredSdk);
     }
 
@@ -140,7 +142,7 @@ public class AgentInstaller {
       agentBuilder = agentBuilder.with(new ExposeAgentBootstrapListener(inst));
     }
 
-    agentBuilder = configureIgnoredTypes(sdkConfig, agentBuilder);
+    agentBuilder = configureIgnoredTypes(sdkConfig, extensionClassLoader, agentBuilder);
 
     if (AgentConfig.isDebugModeEnabled(sdkConfig)) {
       agentBuilder =
@@ -152,7 +154,7 @@ public class AgentInstaller {
     }
 
     int numberOfLoadedExtensions = 0;
-    for (AgentExtension agentExtension : loadOrdered(AgentExtension.class)) {
+    for (AgentExtension agentExtension : loadOrdered(AgentExtension.class, extensionClassLoader)) {
       if (logger.isLoggable(FINE)) {
         logger.log(
             FINE,
@@ -196,9 +198,11 @@ public class AgentInstaller {
     }
   }
 
-  private static void setBootstrapPackages(ConfigProperties config) {
+  private static void setBootstrapPackages(
+      ConfigProperties config, ClassLoader extensionClassLoader) {
     BootstrapPackagesBuilderImpl builder = new BootstrapPackagesBuilderImpl();
-    for (BootstrapPackagesConfigurer configurer : load(BootstrapPackagesConfigurer.class)) {
+    for (BootstrapPackagesConfigurer configurer :
+        load(BootstrapPackagesConfigurer.class, extensionClassLoader)) {
       configurer.configure(builder, config);
     }
     BootstrapPackagePrefixesHolder.setBoostrapPackagePrefixes(builder.build());
@@ -209,9 +213,10 @@ public class AgentInstaller {
   }
 
   private static AgentBuilder configureIgnoredTypes(
-      ConfigProperties config, AgentBuilder agentBuilder) {
+      ConfigProperties config, ClassLoader extensionClassLoader, AgentBuilder agentBuilder) {
     IgnoredTypesBuilderImpl builder = new IgnoredTypesBuilderImpl();
-    for (IgnoredTypesConfigurer configurer : loadOrdered(IgnoredTypesConfigurer.class)) {
+    for (IgnoredTypesConfigurer configurer :
+        loadOrdered(IgnoredTypesConfigurer.class, extensionClassLoader)) {
       configurer.configure(builder, config);
     }
 

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentStarterImpl.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentStarterImpl.java
@@ -62,18 +62,9 @@ public class AgentStarterImpl implements AgentStarter {
   @Override
   public void start() {
     extensionClassLoader = createExtensionClassLoader(getClass().getClassLoader(), javaagentFile);
-    ClassLoader savedContextClassLoader = Thread.currentThread().getContextClassLoader();
-    try {
-      Thread.currentThread().setContextClassLoader(extensionClassLoader);
-      internalStart();
-    } finally {
-      Thread.currentThread().setContextClassLoader(savedContextClassLoader);
-    }
-  }
 
-  private void internalStart() {
     Iterator<LoggingCustomizer> loggingCustomizers =
-        ServiceLoader.load(LoggingCustomizer.class).iterator();
+        ServiceLoader.load(LoggingCustomizer.class, extensionClassLoader).iterator();
     LoggingCustomizer loggingCustomizer;
     if (loggingCustomizers.hasNext()) {
       loggingCustomizer = loggingCustomizers.next();
@@ -84,7 +75,7 @@ public class AgentStarterImpl implements AgentStarter {
     Throwable startupError = null;
     try {
       loggingCustomizer.init();
-      AgentInstaller.installBytebuddyAgent(instrumentation);
+      AgentInstaller.installBytebuddyAgent(instrumentation, extensionClassLoader);
       WeakConcurrentMapCleaner.start();
     } catch (Throwable t) {
       // this is logged below and not rethrown to avoid logging it twice

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/OpenTelemetryInstaller.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/OpenTelemetryInstaller.java
@@ -5,11 +5,9 @@
 
 package io.opentelemetry.javaagent.tooling;
 
-import io.opentelemetry.javaagent.bootstrap.AgentInitializer;
 import io.opentelemetry.javaagent.bootstrap.OpenTelemetrySdkAccess;
 import io.opentelemetry.sdk.OpenTelemetrySdk;
 import io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdk;
-import io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdkBuilder;
 import io.opentelemetry.sdk.common.CompletableResultCode;
 import java.util.Arrays;
 
@@ -21,17 +19,14 @@ public final class OpenTelemetryInstaller {
    *
    * @return the {@link AutoConfiguredOpenTelemetrySdk}
    */
-  public static AutoConfiguredOpenTelemetrySdk installOpenTelemetrySdk() {
-    AutoConfiguredOpenTelemetrySdkBuilder builder =
-        AutoConfiguredOpenTelemetrySdk.builder().setResultAsGlobal(true);
+  public static AutoConfiguredOpenTelemetrySdk installOpenTelemetrySdk(
+      ClassLoader extensionClassLoader) {
 
-    ClassLoader classLoader = AgentInitializer.getExtensionsClassLoader();
-    if (classLoader != null) {
-      // May be null in unit tests.
-      builder.setServiceClassLoader(classLoader);
-    }
-
-    AutoConfiguredOpenTelemetrySdk autoConfiguredSdk = builder.build();
+    AutoConfiguredOpenTelemetrySdk autoConfiguredSdk =
+        AutoConfiguredOpenTelemetrySdk.builder()
+            .setResultAsGlobal(true)
+            .setServiceClassLoader(extensionClassLoader)
+            .build();
     OpenTelemetrySdk sdk = autoConfiguredSdk.getOpenTelemetrySdk();
 
     OpenTelemetrySdkAccess.internalSetForceFlush(

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/SafeServiceLoader.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/SafeServiceLoader.java
@@ -30,9 +30,9 @@ public final class SafeServiceLoader {
    */
   // Because we want to catch exception per iteration
   @SuppressWarnings("ForEachIterable")
-  public static <T> List<T> load(Class<T> serviceClass) {
+  public static <T> List<T> load(Class<T> serviceClass, ClassLoader classLoader) {
     List<T> result = new ArrayList<>();
-    ServiceLoader<T> services = ServiceLoader.load(serviceClass);
+    ServiceLoader<T> services = ServiceLoader.load(serviceClass, classLoader);
     for (Iterator<T> iterator = new SafeIterator<>(services.iterator()); iterator.hasNext(); ) {
       T service = iterator.next();
       if (service != null) {
@@ -43,11 +43,12 @@ public final class SafeServiceLoader {
   }
 
   /**
-   * Same as {@link #load(Class)}, but also orders the returned implementations by comparing their
-   * {@link Ordered#order()}.
+   * Same as {@link #load(Class, ClassLoader)}, but also orders the returned implementations by
+   * comparing their {@link Ordered#order()}.
    */
-  public static <T extends Ordered> List<T> loadOrdered(Class<T> serviceClass) {
-    List<T> result = load(serviceClass);
+  public static <T extends Ordered> List<T> loadOrdered(
+      Class<T> serviceClass, ClassLoader classLoader) {
+    List<T> result = load(serviceClass, classLoader);
     result.sort(Comparator.comparing(Ordered::order));
     return result;
   }

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/InstrumentationLoader.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/InstrumentationLoader.java
@@ -13,6 +13,7 @@ import com.google.auto.service.AutoService;
 import io.opentelemetry.javaagent.bootstrap.InstrumentationHolder;
 import io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule;
 import io.opentelemetry.javaagent.tooling.AgentExtension;
+import io.opentelemetry.javaagent.tooling.Utils;
 import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
 import java.util.logging.Logger;
 import net.bytebuddy.agent.builder.AgentBuilder;
@@ -27,7 +28,8 @@ public class InstrumentationLoader implements AgentExtension {
   @Override
   public AgentBuilder extend(AgentBuilder agentBuilder, ConfigProperties config) {
     int numberOfLoadedModules = 0;
-    for (InstrumentationModule instrumentationModule : loadOrdered(InstrumentationModule.class)) {
+    for (InstrumentationModule instrumentationModule :
+        loadOrdered(InstrumentationModule.class, Utils.getExtensionsClassLoader())) {
       if (logger.isLoggable(FINE)) {
         logger.log(
             FINE,

--- a/javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/test/HelperInjectionTest.groovy
+++ b/javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/test/HelperInjectionTest.groovy
@@ -61,7 +61,7 @@ class HelperInjectionTest extends Specification {
   def "helpers injected on bootstrap classloader"() {
     setup:
     ByteBuddyAgent.install()
-    AgentInstaller.installBytebuddyAgent(ByteBuddyAgent.getInstrumentation())
+    AgentInstaller.installBytebuddyAgent(ByteBuddyAgent.getInstrumentation(), this.class.classLoader)
     String helperClassName = HelperInjectionTest.getPackage().getName() + '.HelperClass'
     HelperInjector injector = new HelperInjector("test", [helperClassName], [], this.class.classLoader, ByteBuddyAgent.getInstrumentation())
     URLClassLoader bootstrapChild = new URLClassLoader(new URL[0], (ClassLoader) null)

--- a/javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/OpenTelemetryInstallerTest.groovy
+++ b/javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/OpenTelemetryInstallerTest.groovy
@@ -24,7 +24,7 @@ class OpenTelemetryInstallerTest extends Specification {
 
   def "should initialize GlobalOpenTelemetry"() {
     when:
-    def otelInstaller = OpenTelemetryInstaller.installOpenTelemetrySdk()
+    def otelInstaller = OpenTelemetryInstaller.installOpenTelemetrySdk(OpenTelemetryInstaller.classLoader)
 
     then:
     otelInstaller != null

--- a/javaagent-tooling/src/test/java/io/opentelemetry/javaagent/tooling/config/ConfigurationFileLoaderTest.java
+++ b/javaagent-tooling/src/test/java/io/opentelemetry/javaagent/tooling/config/ConfigurationFileLoaderTest.java
@@ -46,7 +46,7 @@ class ConfigurationFileLoaderTest {
 
     // when
     AutoConfiguredOpenTelemetrySdk autoConfiguredSdk =
-        OpenTelemetryInstaller.installOpenTelemetrySdk();
+        OpenTelemetryInstaller.installOpenTelemetrySdk(this.getClass().getClassLoader());
 
     // then
     assertThat(autoConfiguredSdk.getConfig().getString("custom.key")).isEqualTo("42");

--- a/muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentTooling.java
+++ b/muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentTooling.java
@@ -32,7 +32,8 @@ public final class AgentTooling {
 
   private static ClassLoader getBootstrapProxy() {
     Iterator<BootstrapProxyProvider> iterator =
-        ServiceLoader.load(BootstrapProxyProvider.class).iterator();
+        ServiceLoader.load(BootstrapProxyProvider.class, AgentTooling.class.getClassLoader())
+            .iterator();
     if (iterator.hasNext()) {
       BootstrapProxyProvider bootstrapProxyProvider = iterator.next();
       return bootstrapProxyProvider.getBootstrapProxy();

--- a/muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/ClassLoaderMatcher.java
+++ b/muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/ClassLoaderMatcher.java
@@ -33,7 +33,7 @@ public class ClassLoaderMatcher {
   public static Map<String, List<Mismatch>> matchesAll(
       ClassLoader classLoader, boolean injectHelpers, Set<String> excludedInstrumentationNames) {
     Map<String, List<Mismatch>> result = new HashMap<>();
-    ServiceLoader.load(InstrumentationModule.class)
+    ServiceLoader.load(InstrumentationModule.class, ClassLoaderMatcher.class.getClassLoader())
         .forEach(
             module -> {
               if (module.instrumentationNames().stream()

--- a/muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/ReferencesPrinter.java
+++ b/muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/ReferencesPrinter.java
@@ -26,7 +26,7 @@ public final class ReferencesPrinter {
    */
   public static void printMuzzleReferences() {
     for (InstrumentationModule instrumentationModule :
-        ServiceLoader.load(InstrumentationModule.class)) {
+        ServiceLoader.load(InstrumentationModule.class, ReferencesPrinter.class.getClassLoader())) {
       try {
         System.out.println(instrumentationModule.getClass().getName());
         for (ClassRef ref :

--- a/testing/agent-exporter/src/main/java/io/opentelemetry/javaagent/testing/bytebuddy/TestAgentListener.java
+++ b/testing/agent-exporter/src/main/java/io/opentelemetry/javaagent/testing/bytebuddy/TestAgentListener.java
@@ -10,6 +10,7 @@ import static java.util.logging.Level.SEVERE;
 import io.opentelemetry.javaagent.extension.ignore.IgnoredTypesConfigurer;
 import io.opentelemetry.javaagent.tooling.EmptyConfigProperties;
 import io.opentelemetry.javaagent.tooling.SafeServiceLoader;
+import io.opentelemetry.javaagent.tooling.Utils;
 import io.opentelemetry.javaagent.tooling.ignore.AdditionalLibraryIgnoredTypesConfigurer;
 import io.opentelemetry.javaagent.tooling.ignore.GlobalIgnoredTypesConfigurer;
 import io.opentelemetry.javaagent.tooling.ignore.IgnoreAllow;
@@ -51,7 +52,8 @@ public class TestAgentListener implements AgentBuilder.Listener {
   private static Trie<IgnoreAllow> buildOtherConfiguredIgnores() {
     IgnoredTypesBuilderImpl builder = new IgnoredTypesBuilderImpl();
     for (IgnoredTypesConfigurer configurer :
-        SafeServiceLoader.loadOrdered(IgnoredTypesConfigurer.class)) {
+        SafeServiceLoader.loadOrdered(
+            IgnoredTypesConfigurer.class, Utils.getExtensionsClassLoader())) {
       // skip built-in agent ignores
       if (configurer instanceof AdditionalLibraryIgnoredTypesConfigurer
           || configurer instanceof GlobalIgnoredTypesConfigurer) {

